HW1/P5/P5_helper.py

```diff
@@ -0,0 +1,82 @@
+from pyspark.accumulators import AccumulatorParam
+class BFSAccumulatorParam(AccumulatorParam):
+	"""
+	Defines an accumulator for the optimized search. The 
+	accumulator is a dict with key: Character, val: distances
+	"""
+	def zero(self, value):
+		return {}
+	def addInPlace(self, val1, val2):
+		# A way to merge two dicts and keep the min of conflicts. 
+		ret = {}
+		for key in list(val1.keys()):
+			if key not in val2:
+				ret[key] = val1[key]
+			else:
+				ret[key] = min(val1[key], val2[key], key=lambda (distance, parent): distance)
+		for key in list(val2.keys()):
+			if key not in val1:
+				ret[key] = val2[key]
+		return ret
+
+def find_neighbors(adjacencyList, node):
+	new_rdd = adjacencyList.filter(lambda k, v: k == node)
+	assert len(new_rdd == 1)
+	return new_rdd.map(lambda k, v: v).take
+
+def expand_node(frontier_distance):
+	def helper((node, (distance, parent, neighbors))):
+		if distance == frontier_distance:
+			print "Found one"
+			return [(new_node, (distance + 1, node, [])) for new_node in neighbors]
+		else:
+			return [(node, (distance, parent, neighbors))]
+	return helper
+def combine_nodes((node, distances)):
+	ret = min(list(distances), key = lambda (distance, parent, neighbors) : distance)
+	neighbors = max(list(distances), key = lambda (distance, parent, neighbors) : len(neighbors))
+	return (node, (ret[0], ret[1], neighbors[2]))
```

Comment:
I guess this should be the most common and efficient way of building the graph and then finding path. Is this the "bfs" for paralleling computing?